### PR TITLE
Add support for custom table filters

### PR DESCRIPTION
**Pull request by vojtechszocs**
`Table` and `List` components now support custom filter functions via existing `ListPage.rowFilters` prop pass-through.

For example:

```ts
// packages/kubevirt-plugin/src/components/table-filters.ts

import { TableFilter } from '@console/internal/components/factory/table-filters';

export const kubevirtTableFilters: TableFilters = {
  VmStatus: {
    type: 'vm-status',
    filter: (statuses, vm) => {
      const status = getSimpleVmStatus(vm);
      return statuses.selected.has(status) || !_.includes(statuses.all, status);
    },
  },
};

type TableFilters = {
  [name: string]: {
    type: string;
    filter: TableFilter;
  };
};
```

```tsx
// packages/kubevirt-plugin/src/components/vm.tsx

import { kubevirtTableFilters } from './table-filters';

const filters = [{
  type: kubevirtTableFilters.VmStatus.type, // change here
  selected: VM_SIMPLE_STATUS_ALL,
  reducer: getSimpleVmStatus,
  items: VM_SIMPLE_STATUS_ALL.map((status) => ({
    id: status,
    title: VM_SIMPLE_STATUS_TO_TEXT[status],
  })),
  filter: kubevirtTableFilters.VmStatus.filter, // change here
}];

// declare VirtualMachinesPage as usual, passing rowFilters={filters}
```

Additionally, `Table` vs. `List` filter map inconsistency was removed by using the same `tableFilters` implementation.

---

/cc @spadgett @christianvogt @mareklibra @jtomasek

### CODE REVIEW

**Comment by vojtechszocs** [comment]
FYI @jelkosz

**Comment by vojtechszocs** [comment]
@spadgett @alecmerdler Can you please share your knowledge on `rowFilters` prop of `ListPage` component?

For example, in `public/components/pod.tsx`

```ts
const filters = [{
  type: 'pod-status',
  selected: [ 'Running', 'Pending', 'Terminating', 'CrashLoopBackOff' ],
  reducer: podPhaseFilterReducer,
  items: [
    { id: 'Running', title: 'Running' },
    { id: 'Pending', title: 'Pending' },
    { id: 'Terminating', title: 'Terminating' },
    { id: 'CrashLoopBackOff', title: 'CrashLoopBackOff' },
    // Use title "Completed" to match what appears in the status column for the pod.
    // The pod phase is "Succeeded," but the container state is "Completed."
    { id: 'Succeeded', title: 'Completed' },
    { id: 'Failed', title: 'Failed' },
    { id: 'Unknown', title: 'Unknown '},
  ],
}];
```

The `rowFilters` prop is used to construct `RowsOfRowFilters` in `ListPageWrapper_` representing the various (visual) filter components, and is therefore unrelated to what we're trying to do in this PR, right?

**Comment by christianvogt** [comment]
As discussed, I do not understand why `rowFilters` is disconnected from the filter definitions which are currently mapped by type inside `table.tsx` and `list.tsx`. Why the decoupling? 
Could we not add a `filter: (selected, all) => boolean` property to the filter definition objects?

**Comment by spadgett** [comment]
> As discussed, I do not understand why `rowFilters` is disconnected from the filter function declarations which are currently mapped by `type` string inside `table.tsx` and `list.tsx`. Why the decoupling?
> Could we not add a `filter: (selected, all) => boolean` property to the filter definition objects?

I agree, that's a cleaner approach. I don't know the history of the current implementation. @alecmerdler might.

We should consider the same thing for `sortFunc`, which follows the same pattern.

**Comment by vojtechszocs** [comment]
PR updated, added KubeVirt example code in PR description.

File: frontend/public/components/factory/table.tsx [hunk 1]
```diff
@@ -52,152 +39,27 @@ import {
...
+const rowFiltersToFilterFuncs = (rowFilters) => {
+  return rowFilters
+    .filter(f => f.type && _.isFunction(f.filter))
+    .reduce((acc, f) => ({ ...acc, [f.type]: f.filter }), {});
 };
 
-const getFilteredRows = (_filters, objects) => {
+const getAllTableFilters = (rowFilters) => ({
+  ...rowFiltersToFilterFuncs(rowFilters),
+  ...tableFilters,
```
Comment (christianvogt) [inline frontend/public/components/factory/table.tsx]: Switch the two around so that rowFilters takes priority?

Comment (vojtechszocs) [reply frontend/public/components/factory/table.tsx]: Agreed, will do.

**Comment by mareklibra** [comment]
@spadgett , @christianvogt , @vojtechszocs , is there a way how to pass additional objects to row filters? As an example, to determine status of a virtual machine, additional objects (like pods or custom resource objects) need to be inspected.

**Comment by vojtechszocs** [comment]
> @spadgett , @christianvogt , @vojtechszocs , is there a way how to pass additional objects to row filters? As an example, to determine status of a virtual machine, additional objects (like pods or custom resource objects) need to be inspected.

@mareklibra This can be done as a follow-up. It seems that the VM table needs more data beyond the usual list of VM objects, so one way to solve that is to make the VM table work with composite (synthetic) objects that combine VM data and any other k8s resource data that is necessary.

**Pull request merged**